### PR TITLE
feat(config): `snapshotPathTemplate` as optional funtion

### LINKED ISSUE #33098: [Feature]: `snapshotPathTemplate` as optional function
### 🚀 Feature Request

I wish to support the following feature in config:

```
snapshotPathTemplate?: string | ((testInfo: TestInfo) => string)
```
OR
```
snapshotPathResolver?: (filename: string, testInfo: TestInfo) => string
```

`snapshotPathResolver` should be invoked with the parsed snapshot path (=`filename`) for further handling, returning the final path to be used.

Currently I am patching playwright to support this.
As you can see it is a very simple change.

```diff

diff --git a/node_modules/playwright/.DS_Store b/node_modules/playwright/.DS_Store
new file mode 100644
index 0000000..5008ddf
Binary files /dev/null and b/node_modules/playwright/.DS_Store differ
diff --git a/node_modules/playwright/lib/common/config.js b/node_modules/playwright/lib/common/config.js
index 4b21d26..690c308 100644
--- a/node_modules/playwright/lib/common/config.js
+++ b/node_modules/playwright/lib/common/config.js
@@ -158,6 +158,7 @@ class FullProjectInternal {
     const testDir = takeFirst(pathResolve(configDir, projectConfig.testDir), pathResolve(configDir, config.testDir), fullConfig.configDir);
     const defaultSnapshotPathTemplate = '{snapshotDir}/{testFileDir}/{testFileName}-snapshots/{arg}{-projectName}{-snapshotSuffix}{ext}';
     this.snapshotPathTemplate = takeFirst(projectConfig.snapshotPathTemplate, config.snapshotPathTemplate, defaultSnapshotPathTemplate);
+    this.snapshotPathResolver = takeFirst(projectConfig.snapshotPathResolver, config.snapshotPathResolver, p => p);
     this.project = {
       grep: takeFirst(projectConfig.grep, config.grep, defaultGrep),
       grepInvert: takeFirst(projectConfig.grepInvert, config.grepInvert, null),
diff --git a/node_modules/playwright/lib/worker/testInfo.js b/node_modules/playwright/lib/worker/testInfo.js
index 5ed9668..61bce7f 100644
--- a/node_modules/playwright/lib/worker/testInfo.js
+++ b/node_modules/playwright/lib/worker/testInfo.js
@@ -386,7 +386,8 @@ class TestInfoImpl {
     const parsedRelativeTestFilePath = _path.default.parse(relativeTestFilePath);
     const projectNamePathSegment = (0, _utils.sanitizeForFilePath)(this.project.name);
     const snapshotPath = (this._projectInternal.snapshotPathTemplate || '').replace(/\{(.)?testDir\}/g, '$1' + this.project.testDir).replace(/\{(.)?snapshotDir\}/g, '$1' + this.project.snapshotDir).replace(/\{(.)?snapshotSuffix\}/g, this.snapshotSuffix ? '$1' + this.snapshotSuffix : '').replace(/\{(.)?testFileDir\}/g, '$1' + parsedRelativeTestFilePath.dir).replace(/\{(.)?platform\}/g, '$1' + process.platform).replace(/\{(.)?projectName\}/g, projectNamePathSegment ? '$1' + projectNamePathSegment : '').replace(/\{(.)?testName\}/g, '$1' + this._fsSanitizedTestName()).replace(/\{(.)?testFileName\}/g, '$1' + parsedRelativeTestFilePath.base).replace(/\{(.)?testFilePath\}/g, '$1' + relativeTestFilePath).replace(/\{(.)?arg\}/g, '$1' + _path.default.join(parsedSubPath.dir, parsedSubPath.name)).replace(/\{(.)?ext\}/g, parsedSubPath.ext ? '$1' + parsedSubPath.ext : '');
-    return _path.default.normalize(_path.default.resolve(this._configInternal.configDir, snapshotPath));
+    const resolvedSnapshotPath = this._projectInternal.snapshotPathResolver ? this._projectInternal.snapshotPathResolver(snapshotPath, this) : snapshotPath;
+    return _path.default.normalize(_path.default.resolve(this._configInternal.configDir, resolvedSnapshotPath));
   }
   skip(...args) {
     this._modifier('skip', args);


```

### Example

It should be used for fine grained runtime control over the snapshot path.

```
  snapshotPathResolver: (file: string, testInfo: TestInfo) => {
    if (CI) {
      return file;
    }

    const { name, ext, dir } = path.parse(file);
    const resolved = path.join(dir, `${name}.${process.platform}${ext}`);
    if (existsSync(path.resolve(resolved))) {
      console.warn(`Using local snapshot ${resolved}`);
      return resolved;
    }
    return file;
  }

```

### Motivation

In a project I work on some snapshots fail on CI (a very little number) due to OS diffs and we do not want to handle 2 sets of snapshot for dev and for CI so I would like to snapshot locally against a different snapshot

### PR DESCRIPTION
**Pull request by ShaMan123**
closes #33098
replaces #33420

This PR adds the possibility to pass a function to `snapshotPathTemplate`, allowing full **runtime** control over snapshot names.

```typescript
    // playwright.config.ts
    import { defineConfig } from '@playwright/test';

    export default defineConfig({
      testDir: './tests',
      snapshotPathTemplate: ({ testDir, testFilePath, arg, ext }) => 
            `${testDir}/__screenshots__/${testFilePath}/${arg}${ext}`,
    });

```

I am not sure the name fits this functionality so it might be better to add this as a standalone option.

### CODE REVIEW

File: docs/src/api/params.md [hunk 1]
```diff
@@ -1751,7 +1751,7 @@ await Expect(Page.GetByTitle("Issues count")).toHaveText("25 issues");
 ```
 
 ## test-config-snapshot-path-template
-- `type` ?<[string]>
+- `type` ?<[string]|[SnapshotPathResolver]>
```
Comment (ShaMan123) [inline docs/src/api/params.md:1754]: How do I import `SnapshotPathResolver` into the generated .d.ts file?

**Review by ShaMan123 — COMMENTED** [review]
functionality is stable, docs aren't

**Comment by ShaMan123** [comment]
@microsoft-github-policy-service agree

**Comment by pavelfeldman** [comment]
Not seeing consensus in the issue that would allow for a PR.

**Comment by ShaMan123** [comment]
You did see with #33420?

**Comment by ShaMan123** [comment]
Because that was going to be merged

**Comment by ShaMan123** [comment]
@dgozman @sisha

**Comment by ShaMan123** [comment]
please reopen => https://github.com/Azure/azure-sdk-for-js/pull/35996

**Pull request closed without merging**